Fix #5460: Improve completion of import nodes

Duhemm opened the pull request.

Duhemm commented:
Ah that's a good point, this is indeed broken at the moment

I'm trying to avoid duplicates when doing `import java.io.FileDescrip<tab>`. This condition should be checked only if we're completing imports.

smarter commented on compiler/src/dotty/tools/dotc/interactive/Interactive.scala:
```diff
@@ -166,31 +166,72 @@ object Interactive {
+     *   3. are not a primary constructor,
+     *   4. have an existing source symbol,
+     *   5. are the module class in case of packages,
+     *   6. are not Java module classes when completing imports (to avoid duplicate results).
```
Don't you also get duplicates if there's both a Scala class and object with the same name ?

Duhemm replied:
Yes, but only if there explicitly exists a class and an object, so I personally find it less disturbing

https://github.com/lampepfl/dotty/blob/eba580ccf17c37cde6562beefad5e7f0b0345029/language-server/test/dotty/tools/languageserver/CompletionTest.scala#L83-L91

I'm happy to just remove this commit if you think it's not worth it. Or we could bypass the problem by removing all duplicate names from the completion lists. Or keep the duplicates.

smarter replied:
> Or we could bypass the problem by removing all duplicate names from the completion lists. 

Yeah I would just do that, if there's both a class and an object, then just show the symbol for class in the completion

smarter replied:
Also ideally in term position we should just show the objects, and in type position the class.

smarter commented on compiler/src/dotty/tools/dotc/interactive/Interactive.scala:
```diff
@@ -166,31 +166,72 @@ object Interactive {
+     * @return The point where to insert completion, whether terms should be included in results,
+     *         whether types should be included, and whether we're completing an import.
+     */
+    def completionInfo(path: List[Tree], inImport: Boolean): (Int, String, Boolean, Boolean, Boolean) = path match {
```
That's a lot of stuff you're returning, this should either be a case class or the code should be refactored to avoid having to return things at this point

smarter commented on language-server/test/dotty/tools/languageserver/CompletionTest.scala:
```diff
@@ -1,22 +1,146 @@
+      code"""object Foo { class MyClass }""",
+      code"""import Foo.${m1}"""
+    ).completion(m1, results => {
+      val myClass = ("MyClass", Class, "Object{...}")
```
This "Object{...}" thing is weird, we should either not display anything, or list the class parents.

Duhemm replied:
I've changed it so that we show the type for terms (like we do currently), and the fully qualified name for types. What do you think?

smarter commented on language-server/src/dotty/tools/languageserver/DottyLanguageServer.scala:
```diff
@@ -794,12 +794,16 @@ object DottyLanguageServer {
+    item.setDetail(detail)
+    ParsedComment.docOf(sym).foreach { doc =>
+      item.setDocumentation(markupContent(doc.renderAsMarkdown))
+    }
```
While you're extending what we show in completion: CompletionItem has a `deprecated` field we could set now.

Duhemm replied:
Good idea. That's done, but I can't see a different in VSCode. The tests show it is actually marked as deprecated, though...

Duhemm commented on compiler/src/dotty/tools/dotc/interactive/Interactive.scala:
```diff
@@ -166,33 +166,76 @@ object Interactive {
-      (!termOnly || sym.isTerm) &&
-      (!typeOnly || sym.isType)
+      sym.name.startsWith(info.prefix) &&
+      !sym.name.toString.drop(info.prefix.length).contains('$') &&
```
I think that the intention is to keep names with special characters if the prefix was explicitly written by the user.

Duhemm commented:
@smarter I've moved everything related to completions out of `object Interactive` and into `object Completion`. I've refactored it to make it easier to follow, as before everything was crammed in one method.

The actual logic of all the different steps needed to get all the completion candidates is unchanged, though.

I've removed `termOnly` and `typeOnly` in favor of a `Mode` which can be either `Term`, `Type`, `Import` or `None`. `Mode.None` will indicate that we cannot provide any suggestion for completions (for instance, trying to autocomplete a renaming import), so we don't have to worry about trying to complete at `offset = 0`.

smarter commented:
> Mode.None will indicate that we cannot provide any suggestion for completions (for instance, trying to autocomplete a renaming import)

Could we have an `Option[CompletionMode]` instead, since in the `None` case we could just early return with an empty list ?

Duhemm commented:
> Could we have an Option[CompletionMode] instead, since in the None case we could just early return with an empty list ?

That's already the case. I don't think it'd be cleaner to use an `Option`, but I'm happy to change it if you prefer.

Duhemm commented:
@smarter More comments on this PR?

Duhemm commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -0,0 +1,307 @@
+          addMember(imp.site, name)
+          addMember(imp.site, name.toTypeName)
+        }
+        // FIXME: We need to also take renamed items into account for completions,
```
Yes, see #3979, and an open PR: #4977

Duhemm commented:
@smarter I've added `_` to the list of completions when in import mode and `qual.tpe.isStable`. It looks like this is the check that is when typing the import. I don't know whether it should check realizability, nor why the check is skipped after typer :

https://github.com/lampepfl/dotty/blob/30b8b1ce174a2d0a2d05e5106c862d3c499c7908/compiler/src/dotty/tools/dotc/typer/Typer.scala#L1717-L1722

https://github.com/lampepfl/dotty/blob/30b8b1ce174a2d0a2d05e5106c862d3c499c7908/compiler/src/dotty/tools/dotc/typer/Checking.scala#L580-L582

smarter commented on language-server/src/dotty/tools/languageserver/DottyLanguageServer.scala:
```diff
@@ -795,12 +796,17 @@ object DottyLanguageServer {
     val label = sym.name.show
     val item = new lsp4j.CompletionItem(label)
-    item.setDetail(sym.info.widenTermRefExpr.show)
+    val detail = if (sym.isType) sym.showFullName else sym.info.widenTermRefExpr.show
```
Idea: instead of just showing the full name we could show the kind of the symbol too, e.g. "class a.b.c.Foo", "val x", ... and in the case where there's more than one symbol with the same name, we could show both at the same time "class and object a.b.c.Foo", ... We can reuse the code used in the REPL when printing a declaration.

Duhemm commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -45,7 +45,7 @@ object Completion {
     path match {
       case (ref: RefTree) :: _ =>
-        if (ref.name == nme.ERROR) Mode.None
+        if (ref.name == nme.ERROR) Mode.Term | Mode.Type
```
@smarter 

> Is this because we don't know if we're completing a term or a type at this point ? The parser should know if we're in term or type context, and could output nme.ERROR or tpnme.ERROR depending on which one it is.

Good idea!

> Also, thinking about it, when completing something in type context we should also complete terms, because a term can have a type member (but a type cannot have a term member)

I've added that too. From the spec, it looks like the path for types needs to include stable members only. I've used `CheckRealizability` for that, but I'm not sure this is correct.

smarter commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -219,9 +220,20 @@ object Completion {
+    /** Can this symbol be part of a path? See SLS 3.1 for a definition of a valid path. */
+    private def validPathSegment(sym: Symbol)(implicit ctx: Context): Boolean = {
+      def isRealizable = {
+        val realizability = CheckRealizable.realizability(sym.info)
```
I'm afraid that could be a bit slow, I think just checking isStable should be good enough

allanrenucci commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -45,7 +45,8 @@ object Completion {
       case (ref: RefTree) :: _ =>
-        if (ref.name == nme.ERROR) Mode.Term | Mode.Type
+        if (ref.name == tpnme.ERROR) Mode.Type
+        else if (ref.name == nme.ERROR) Mode.Term
```
I don't think you need to special case `ERROR`:
```scala
nme.ERROR.isTermName   ==> true
nme.ERROR.isTypeName   ==> false
tpnme.ERROR.isTermName ==> false
tpnme.ERROR.isTypeName ==> true

```

smarter commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -219,9 +220,20 @@ object Completion {
       !sym.is(allOf(Mutable, Accessor)) &&
       (
            (mode.is(Mode.Term) && sym.isTerm)
+        || (mode.is(Mode.StableTerm) && sym.isTerm && validPathSegment(sym))
```
I don't think adding a mode is necessary, you could do `mode.is(Mode.Type) && (sym.isType || sym.isStable)` I think

smarter commented:
Doesn't have to be done now but I think what we should do eventually is make our own custom CompletionItem data type and return a `List[CompletionItem]` instead of a `List[Symbol], it could be defined a bit like this:
```scala
case class Completion(label: String, description: String, documentation: String, symbols: List[String])
```
This way we can handle completions with no symbol (wildcards, structural types) and multiple symbols (both a class and an object exist, ...) in a uniform way. For example when there's both a class and an object, we could have the `documentation` field contain the documentation for both.
